lib/collection/src/collection.rs

```diff
@@ -258,17 +258,49 @@ impl Collection {
         todo!("Handle for sender and receiver")
     }
 
+    pub async fn initiate_temporary_shard(&self, shard_id: ShardId) -> CollectionResult<()> {
+        let mut shards_holder = self.shards_holder.write().await;
+        shards_holder
+            .initiate_temporary_shard(self.id.clone(), &self.path, self.config.clone(), shard_id)
+            .await
+    }
+
+    /// Handle collection updates from peers.
+    ///
+    /// Shard transfer aware.
     pub async fn update_from_peer(
         &self,
         operation: CollectionUpdateOperations,
         shard_selection: ShardId,
         wait: bool,
     ) -> CollectionResult<UpdateResult> {
-        let local_shard = self
-            .shards_holder
-            .local_shard_by_id(shard_selection)
-            .await?;
-        local_shard.get().update(operation.clone(), wait).await
+        let shard_opt = self.shards_holder.get_shard(shard_selection).await;
+
+        let target_shard = match shard_opt {
+            None => {
+                return Err(CollectionError::bad_shard_selection(format!(
+                    "Shard {} does not exist",
+                    shard_selection
+                )))
+            }
+            Some(shard) => match *shard {
+                Shard::Local(_) => shard,
+                Shard::Proxy(_) => shard,
+                Shard::Remote(_) => {
+                    // check temporary shards if the target is a remote shard
+                    let temporary_shard_opt = self
+                        .shards_holder
+                        .get_temporary_shard(shard_selection)
```

Comment:
it looks like a double read lock to me: `get_shard` already acquired the lock for `shards_holder`

Reply:
Good catch :+1:

Reply:
https://github.com/qdrant/qdrant/pull/829/commits/a54803b28fd7e2b811352ac33b91a1371c375a4d